dev/PowerNotifications/ClosedSourceMock.cpp

```diff
@@ -0,0 +1,271 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include <pch.h>
+#include <unordered_map>
+#include <PowerManager.g.h>
+#include <PowerNotificationsPal.h>
+
+using namespace std;
+using index = ULONGLONG;
+index g_idx;
+
+unordered_map<index, OnCompositeBatteryStatusChanged> onCompositeBatteryStatusChanged_callbacks;
+unordered_map<index, OnDischargeTimeChanged> onDischargeTimeChanged_callbacks;
+unordered_map<index, OnEnergySaverStatusChanged> onEnergySaverStatusChanged_callbacks;
+unordered_map<index, OnPowerConditionChanged> onPowerConditionChanged_callbacks;
+unordered_map<index, OnDisplayStatusChanged> onDisplayStatusChanged_callbacks;
+unordered_map<index, OnSystemIdleStatusChanged> onSystemIdleStatusChanged_callbacks;
+unordered_map<index, OnPowerSchemePersonalityChanged> onPowerSchemePersonalityChanged_callbacks;
+unordered_map<index, OnUserPresenceStatusChanged> onUserPresenceStatusChanged_callbacks;
+unordered_map<index, OnSystemAwayModeStatusChanged> onSystemAwayModeStatusChanged_callbacks;
+
+HRESULT GetCompositeBatteryStatus(CompositeBatteryStatus** compositeBatteryStatusOut)
+{
+    *compositeBatteryStatusOut = NULL;
+    auto status = wil::make_unique_cotaskmem<CompositeBatteryStatus>();
+    status.get()->ActiveBatteryCount = 1;
+    status.get()->Status.PowerState |= BATTERY_DISCHARGING;
+    status.get()->Status.PowerState |= BATTERY_POWER_ON_LINE;
+    status.get()->Information.FullChargedCapacity = 100;
+    status.get()->Status.Capacity = 77;
+    *compositeBatteryStatusOut = status.release();
+    return S_OK;
+}
+
+HRESULT RegisterCompositeBatteryStatusChangedListener(OnCompositeBatteryStatusChanged listener, CompositeBatteryStatusRegistration* registration)
+{
+    *registration = reinterpret_cast<CompositeBatteryStatusRegistration>(g_idx++);
+    onCompositeBatteryStatusChanged_callbacks[g_idx] = listener;
+    std::thread thread([]() {
+        std::this_thread::sleep_for(std::chrono::milliseconds(50));
+        for (const auto& [key, callbackFn] : onCompositeBatteryStatusChanged_callbacks)
+        {
+            auto status = wil::make_unique_cotaskmem<CompositeBatteryStatus>();
```

Comment:
With this and the status.release() in the call, you're transferring ownership of the CompositeBatteryStatus to callbackFn. That's not the conventional lifetime semantics for in-params. Normally an in-param is only valid for the duration of the call, and the callee must copy it if they want to persist it; they don't receive ownership, and aren't responsible for freeing the parameter. As-is, if the callee doesn't call CoTaskMemFree at some point (potentially after the call returns), this allocation will leak.

I'd recommend stack-allocating the CompositeBatteryStatus unless we've got a really good reason to break convention.